packages/ckeditor5-clipboard/src/clipboardmarkersutils.ts

```diff
@@ -294,34 +311,103 @@ export default class ClipboardMarkersUtils extends Plugin {
+			// Check if marker exists in configuration and provided action can be performed on it.
+			const isCopyable = this._isMarkerCopyable( marker.name, action );
+
+			if ( !isCopyable ) {
+				return false;
+			}
+
+			// Checks if configuration disallows to copy marker only if part of its content is selected.
+			//
+			// Example:
+			// 	<marker-a> Hello [ World ] </marker-a>
+			//					     ^ selection
+			//
+			// In this scenario `marker-a` won't be copied because selection doesn't overlap its content entirely.
+			const { copyPartiallySelected } = this._getMarkerClipboardConfig( marker.name )!;
+
+			if ( !copyPartiallySelected ) {
+				const markerRange = marker.getRange();
+
+				return selectionRanges.some( selectionRange => selectionRange.containsRange( markerRange, true ) );
+			}
+
+			return true;
+		};
+
 		return Array
-			.from( selection.getRanges()! )
-			.flatMap( selectionRange => Array.from( writer.model.markers.getMarkersIntersectingRange( selectionRange ) ) )
-			.filter( marker => this._canPerformMarkerClipboardAction( marker.name, action ) )
-			.map( ( marker ): CopyableMarker => ( {
-				name: marker.name,
-				range: marker.getRange()
-			} ) );
+			.from( markersInRanges )
+			.filter( isSelectionMarkerCopyable )
+			.map( ( marker ): CopyableMarker => {
+				// During `dragstart` event original marker is still present in tree.
+				// It is removed after the clipboard drop event, so none of the copied markers are inserted at the end.
+				// It happens because there already markers with specified `marker.name` when clipboard is trying to insert data
+				// and it aborts inserting.
+				const name = action === 'dragstart' ? this._getUniqueMarkerName( marker.name ) : marker.name;
+
+				return {
+					name,
+					range: marker.getRange()
+				};
+			} );
 	}
 
 	/**
-	 * Picks all markers from markers map that can be copied.
+	 * Picks all markers from markers map that can be pasted.
+	 * If `duplicateOnPaste` is `true`, it regenerates their IDs to ensure uniqueness.
+	 * If marker is not registered, it will be kept in the array anyway.
 	 *
 	 * @param markers Object that maps marker name to corresponding range.
 	 * @param action Type of clipboard action. If null then checks only if marker is registered as copyable.
 	 */
-	private _getCopyableMarkersFromRangeMap(
+	private _getPasteMarkersFromRangeMap(
 		markers: Record<string, Range> | Map<string, Range>,
 		action: ClipboardMarkerRestrictedAction | null = null
 	): Array<CopyableMarker> {
+		const { model } = this.editor;
 		const entries = markers instanceof Map ? Array.from( markers.entries() ) : Object.entries( markers );
 
-		return entries
-			.map( ( [ markerName, range ] ): CopyableMarker => ( {
-				name: markerName,
-				range
-			} ) )
-			.filter( marker => this._canPerformMarkerClipboardAction( marker.name, action ) );
+		return entries.flatMap( ( [ markerName, range ] ): Array<CopyableMarker> => {
```

Comment:
@DawidKossowski It's a discussion between functional (immutable) and imperative (mutable) approach. From my point of the view, the first one is much easier to debug because it disallows programmer to modify previously added items and allows to easily chain new transformers on already mapped array. In this particular case, performance is not even an issue because it'll work exactly the same for arrays with even ~5000 items.

Reply:
I have nothing against functional programming, but we simply don't use such practices in our project.

@niegowski WDYT?